ReactiveCocoa/NSObject+KeyValueObserving.swift

```diff
@@ -53,6 +53,105 @@ extension Reactive where Base: NSObject {
 			signalLifetime += lifetime.observeEnded(observer.sendCompleted)
 		}
 	}
+
+	private func producer<U>(
+		for keyPath: KeyPath<Base, U>,
+		transform: @escaping (Any?) -> U
+	) -> SignalProducer<U, NoError> {
+		guard let kvcKeyPath = keyPath._kvcKeyPathString else {
+			fatalError("Cannot use `producer(for:)` on a non Objective-C property.")
+		}
+
+		return SignalProducer { observer, lifetime in
+			lifetime += KeyValueObserver.observe(
+				self.base,
+				keyPath: kvcKeyPath,
+				options: [.initial, .new],
+				action: { observer.send(value: transform($0)) }
+			)
+
+			lifetime += self.lifetime.observeEnded(observer.sendCompleted)
+		}
+	}
+
+	private func signal<U>(
+		for keyPath: KeyPath<Base, U>,
+		transform: @escaping (Any?) -> U
+	) -> Signal<U, NoError> {
+		guard let kvcKeyPath = keyPath._kvcKeyPathString else {
+			fatalError("Cannot use `signal(for:)` on a non Objective-C property.")
+		}
+
+		return Signal { observer, lifetime in
+			lifetime += KeyValueObserver.observe(
+				self.base,
+				keyPath: kvcKeyPath,
+				options: [.new],
+				action: { observer.send(value: transform($0)) }
+			)
+			lifetime += self.lifetime.observeEnded(observer.sendCompleted)
+		}
+	}
+
+	/// Create a producer which sends the current value and all the subsequent
+	/// changes of the property specified by the key path.
+	///
+	/// The producer completes when the object deinitializes.
+	///
+	/// - parameters:
+	///   - keyPath: The key path of the property to be observed.
+	///
+	/// - returns: A producer emitting values of the property specified by the
+	///            key path.
+	public func producer<U>(for keyPath: KeyPath<Base, U?>) -> SignalProducer<U?, NoError> {
+		return producer(for: keyPath) { $0 as! U? }
+	}
+
+	/// Create a signal all changes of the property specified by the key path.
+	///
+	/// The signal completes when the object deinitializes.
+	///
+	/// - note:
+	///	  Does not send the initial value. See `producer(forKeyPath:)`.
+	///
+	/// - parameters:
+	///   - keyPath: The key path of the property to be observed.
+	///
+	/// - returns: A producer emitting values of the property specified by the
+	///            key path.
+	public func signal<U>(for keyPath: KeyPath<Base, U?>) -> Signal<U?, NoError> {
+		return signal(for: keyPath) { $0 as! U? }
+	}
```

Comment:
Why are these overloads needed? Can't optionals use the below variants? Won't these introduce ambiguities?

Reply:
> Optionals are special cased, because force casting does not open a generic parameter for optional inspection. So even if U turns out to an optional, coercing Any? to U fails.

Not sure if the runtime casting machinery has changed lately.